File: platform/cli-go/internal/commands/split.go
package commands

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"

	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/elloloop/devpod/platform/cli-go/internal/git"
	"github.com/elloloop/devpod/platform/cli-go/internal/llm"
	"github.com/elloloop/devpod/platform/cli-go/internal/workspace"
	"github.com/spf13/cobra"
)

type splitGroup struct {
	title string
	files []string
}

func askConfirmation(prompt string) bool {
	fmt.Print(prompt)
	scanner := bufio.NewScanner(os.Stdin)
	if scanner.Scan() {
		answer := strings.TrimSpace(strings.ToLower(scanner.Text()))
		return answer == "" || answer == "y" || answer == "yes"
	}
	return false
}

func splitWithLLM(diffContent string, files []string, changeType workspace.ChangeType, scope string) []splitGroup {
	// Group files by directory/module heuristic
	groups := make(map[string][]string)
	var groupOrder []string

	for _, file := range files {
		parts := strings.Split(file, "/")
		var key string
		if len(parts) <= 1 {
			key = "root"
		} else if len(parts) <= 2 {
			key = parts[0]
		} else {
			key = parts[0] + "/" + parts[1]
		}

		if _, exists := groups[key]; !exists {
			groupOrder = append(groupOrder, key)
		}
		groups[key] = append(groups[key], file)
	}

	var result []splitGroup
	for _, area := range groupOrder {
		areaFiles := groups[area]
		var title string

		if llm.IsAvailable() {
			msg, err := llm.GenerateDiffMessage(diffContent, string(changeType), scope)
			if err == nil {
				title = msg.Title
			} else {
				msg := llm.GenerateFallbackMessage(areaFiles, string(changeType), scope)
				title = msg.Title
			}
		} else {
			prefix := format.FeatureTypePrefix(string(changeType))
			title = fmt.Sprintf("%s(%s): update %s", prefix, scope, area)
		}

		result = append(result, splitGroup{title: title, files: areaFiles})
	}

	return result
}

func newSplitCmd() *cobra.Command {
	var preview bool

	cmd := &cobra.Command{
		Use:   "split",
		Short: "Split current changes into multiple diffs",
		RunE: func(cmd *cobra.Command, args []string) error {
			if !llm.IsAvailable() {
				fmt.Println(format.ErrorMsg("Split requires an LLM. Configure with: devpod config set llm.provider anthropic"))
				return fmt.Errorf("LLM not available")
			}

			feature := workspace.GetCurrentFeature()
			if feature == nil {
				fmt.Println(format.ErrorMsg("No active feature."))
				fmt.Println(format.DimText("Start one with: devpod feature \"name\""))
				return fmt.Errorf("no active feature")
			}

			changes, _ := git.GetChangedFiles()
			if len(changes) == 0 {
				fmt.Println(format.DimText("No changes to split."))
				return nil
			}

			diffContent, _ := git.GetDiff()
			changedFiles := make([]string, len(changes))
			for i, c := range changes {
				changedFiles[i] = c.Path
			}
			scope := feature.Slug

			fmt.Println(format.DimText(fmt.Sprintf("Analyzing %d changed files...", len(changedFiles))))

			groups := splitWithLLM(diffContent, changedFiles, feature.Type, scope)

			if len(groups) <= 1 {
				fmt.Println(format.DimText("All changes belong in a single diff. Use \"devpod diff\" instead."))
				return nil
			}

			// Show proposed split
			nextPos := workspace.GetNextDiffPosition(*feature)
			fmt.Println()
			fmt.Printf("Suggested split (%d diffs):\n", len(groups))
			for i, group := range groups {
				label := format.DiffLabel(nextPos + i)
				fileList := strings.Join(group.files, ", ")
				fmt.Printf("  %s: %s \u2014 %s\n", label, group.title, format.DimText(fileList))
			}

			if preview {
				return nil
			}

			fmt.Println()
			if !askConfirmation("Apply? [Y/n] ") {
				fmt.Println(format.DimText("Cancelled."))
				return nil
			}

			// Save undo entry
			headBefore, _ := git.GetHeadSHA()
			_ = workspace.SaveUndoEntry(workspace.UndoEntry{
				Action:      "split",
				Timestamp:   time.Now().UTC().Format(time.RFC3339),
				RefBefore:   headBefore,
				Description: fmt.Sprintf("Split into %d diffs", len(groups)),
				Data:        map[string]interface{}{"count": len(groups)},
			})

			// Apply each group
			for i, group := range groups {
				position := nextPos + i

				// Stage only this group's files
				for _, file := range group.files {
					addCmd := exec.Command("git", "add", file)
					if err := addCmd.Run(); err != nil {
						addCmd2 := exec.Command("git", "add", "-A", file)
						_ = addCmd2.Run()
					}
				}

				// Commit
				commitSHA, _ := git.Commit(group.title)

				// Get stats for this commit
				stats, _ := git.GetDiffStats()

				// Create diff metadata
				uuid := workspace.GenerateDiffUUID()
				newDiff := workspace.DiffData{
					UUID:        uuid,
					Feature:     feature.Slug,
					Commit:      commitSHA,
					Position:    position,
					Title:       group.title,
					Description: "",
					Type:        feature.Type,
					Files:       group.files,
					Additions:   stats.Additions,
					Deletions:   stats.Deletions,
					Version:     1,
					Status:      "draft",
					CI:          "",
					GitHubPR:    0,
					Created:     time.Now().UTC().Format(time.RFC3339),
					Updated:     time.Now().UTC().Format(time.RFC3339),
				}

				_ = workspace.SaveDiff(newDiff)
				feature.Diffs = append(feature.Diffs, uuid)

				fmt.Printf("  %s %s: %s\n", format.SuccessMsg("\u2713"), format.DiffLabel(position), group.title)
			}

			_ = workspace.SaveFeature(*feature)

			fmt.Println()
			fmt.Printf("%s Split into %d diffs\n", format.SuccessMsg("\u2713"), len(groups))
			fmt.Println(format.NextStepHint("split"))
			return nil
		},
	}

	cmd.Flags().BoolVar(&preview, "preview", false, "Preview the split without applying")

	return cmd
}

File: platform/cli/src/commands/split.ts
import { Command } from 'commander';
import chalk from 'chalk';
import * as readline from 'node:readline';
import { execSync } from 'node:child_process';
import * as git from '../git';
import {
  getCurrentFeature,
  getNextDiffPosition,
  generateDiffUuid,
  saveDiff,
  saveFeature,
  saveUndoEntry,
  type DiffData,
  type ChangeType,
} from '../workspace';
import {
  generateDiffMessage,
  isLLMAvailable,
  generateFallbackMessage,
} from '../llm';
import {
  errorMessage,
  diffLabel,
  featureTypePrefix,
  nextStepHint,
} from '../format';

interface SplitGroup {
  title: string;
  files: string[];
}

async function askConfirmation(prompt: string): Promise<boolean> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    rl.question(prompt, (answer) => {
      rl.close();
      const normalized = answer.trim().toLowerCase();
      resolve(normalized === '' || normalized === 'y' || normalized === 'yes');
    });
  });
}

async function splitWithLLM(
  diffContent: string,
  files: string[],
  type: ChangeType,
  scope: string,
): Promise<SplitGroup[]> {
  // For split, we ask the LLM to group files into logical diffs
  // We use generateDiffMessage for each group after splitting
  // The grouping itself is a heuristic based on file paths

  // Simple heuristic grouping by directory/module
  const groups = new Map<string, string[]>();

  for (const file of files) {
    const parts = file.split('/');
    // Group by the first meaningful directory
    let key: string;
    if (parts.length <= 1) {
      key = 'root';
    } else if (parts.length <= 2) {
      key = parts[0]!;
    } else {
      key = `${parts[0]}/${parts[1]}`;
    }

    if (!groups.has(key)) {
      groups.set(key, []);
    }
    groups.get(key)!.push(file);
  }

  // Convert to split groups with titles
  const result: SplitGroup[] = [];
  let position = 1;

  for (const [area, areaFiles] of groups) {
    // Try to generate a title for this group
    let title: string;
    if (isLLMAvailable()) {
      try {
        const msg = await generateDiffMessage(diffContent, type , scope);
        title = msg.title;
      } catch {
        const msg = generateFallbackMessage(areaFiles, type , scope);
        title = msg.title;
      }
    } else {
      const prefix = featureTypePrefix(type);
      title = `${prefix}(${scope}): update ${area}`;
    }

    result.push({ title, files: areaFiles });
    position++;
  }

  return result;
}

export function registerSplitCommand(program: Command): void {
  program
    .command('split')
    .description('Split current changes into multiple diffs')
    .option('--preview', 'Preview the split without applying')
    .action(async (opts: { preview?: boolean }) => {
      try {
        if (!isLLMAvailable()) {
          console.log(
            chalk.red(
              'Split requires an LLM. Configure with: devpod config set llm.provider anthropic',
            ),
          );
          process.exit(1);
          return;
        }

        const feature = getCurrentFeature();
        if (!feature) {
          console.log(chalk.red('No active feature.'));
          console.log(
            chalk.dim('Start one with: devpod feature "name"'),
          );
          process.exit(1);
          return;
        }

        const changes = git.getChangedFiles();
        if (changes.length === 0) {
          console.log(chalk.dim('No changes to split.'));
          return;
        }

        const diffContent = git.getDiff();
        const changedFiles = changes.map((c) => c.path);
        const scope = feature.slug;

        console.log(
          chalk.dim(`Analyzing ${changedFiles.length} changed files...`),
        );

        const groups = await splitWithLLM(
          diffContent,
          changedFiles,
          feature.type,
          scope,
        );

        if (groups.length <= 1) {
          console.log(
            chalk.dim(
              'All changes belong in a single diff. Use "devpod diff" instead.',
            ),
          );
          return;
        }

        // Show proposed split
        const nextPos = getNextDiffPosition(feature);
        console.log();
        console.log(
          chalk.bold(`Suggested split (${groups.length} diffs):`),
        );
        for (let i = 0; i < groups.length; i++) {
          const group = groups[i]!;
          const label = diffLabel(nextPos + i);
          const fileList = group.files.join(', ');
          console.log(
            `  ${label}: ${group.title} \u2014 ${chalk.dim(fileList)}`,
          );
        }

        if (opts.preview) {
          return;
        }

        console.log();
        const confirmed = await askConfirmation('Apply? [Y/n] ');
        if (!confirmed) {
          console.log(chalk.dim('Cancelled.'));
          return;
        }

        // Save undo entry
        const headBefore = git.getHeadSha();
        saveUndoEntry({
          action: 'split',
          timestamp: new Date().toISOString(),
          refBefore: headBefore,
          description: `Split into ${groups.length} diffs`,
          data: { count: groups.length },
        });

        // Apply each group
        for (let i = 0; i < groups.length; i++) {
          const group = groups[i]!;
          const position = nextPos + i;

          // Stage only this group's files
          for (const file of group.files) {
            try {
              execSync(`git add ${JSON.stringify(file)}`, {
                stdio: 'pipe',
                encoding: 'utf-8',
              });
            } catch {
              // File may have been deleted
              try {
                execSync(`git add -A ${JSON.stringify(file)}`, {
                  stdio: 'pipe',
                  encoding: 'utf-8',
                });
              } catch {
                // skip
              }
            }
          }

          // Commit
          const commitSha = git.commit(group.title);

          // Get stats for this commit
          const stats = git.getDiffStats();

          // Create diff metadata
          const uuid = generateDiffUuid();
          const newDiff: DiffData = {
            uuid,
            feature: feature.slug,
            commit: commitSha,
            position,
            title: group.title,
            description: '',
            type: feature.type,
            files: group.files,
            additions: stats.additions,
            deletions: stats.deletions,
            version: 1,
            status: 'draft',
            ci: null,
            githubPr: null,
            created: new Date().toISOString(),
            updated: new Date().toISOString(),
          };

          saveDiff(newDiff);
          feature.diffs.push(uuid);

          console.log(
            `  ${chalk.green('\u2713')} ${diffLabel(position)}: ${group.title}`,
          );
        }

        saveFeature(feature);

        console.log();
        console.log(
          `${chalk.green('\u2713')} Split into ${groups.length} diffs`,
        );
        console.log(nextStepHint('split'));
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}
